lib/Dialect/FIRRTL/Transforms/CheckCombLoops.cpp

```diff
@@ -0,0 +1,580 @@
+  }
+
+  // Local implementation of getFieldRefFromValue, returns the Aggregate type
+  /// Get the FieldRef from a value. Customized implementation of
+  /// getFieldRefFromValue. If `field` is a subfield, then this will return the
+  /// base aggregate Value and the corresponding fieldId. This will return
+  /// nullptr if thie field is a result of a register and can be ignored.
+  FieldRef getFieldRef(Value field) {
+    if (field.isa<BlockArgument>() ||
+        !isa<SubfieldOp, SubaccessOp, SubindexOp>(field.getDefiningOp()))
+      return FieldRef(field, 0);
+    auto iter = leafToBaseValMap.find(field);
+    if (iter == leafToBaseValMap.end())
+      return FieldRef(nullptr, 0);
+    auto baseVal = iter->second;
+    auto storIter = valToStorageLocIdMap.find(baseVal);
+    if (storIter == valToStorageLocIdMap.end())
+      return FieldRef(nullptr, 0);
+    auto baseStorId = storIter->getSecond().front();
+    auto fieldStorIter = valToStorageLocIdMap.find(field);
+    if (fieldStorIter == valToStorageLocIdMap.end())
+      return FieldRef(nullptr, 0);
+    auto fieldStorId = fieldStorIter->getSecond().front();
+
+    return FieldRef(baseVal, fieldStorId - baseStorId);
+  }
+
+  // Add all the Values to `children` that have a valid dataflow from `node`.
+  // The `inputArg` is valid, only if it is one of the reaching definitions to
+  // `node`.
+  void getChildren(Value node, Node &inputArg,
+                   SmallVector<ValueOpPair> &children) {
+
+    // Handle the following cases
+    // Case: Instance port: Add the port fields that have a combinational path
+    // in the referenced module, to the children.
+    //
+    if (!node.isa<BlockArgument>()) {
+      // Any Value that is derived from a Register must be ignored. `skip` is
+      // true, only if this node is the result of some Register.
+      bool skip =
+          TypeSwitch<Operation *, bool>(node.getDefiningOp())
+              .Case<InstanceOp>([&](InstanceOp ins) {
+                size_t portNum = 0;
+                // Get the port index for `node`.
+                for (const auto &p : llvm::enumerate(ins->getOpResults()))
+                  if (p.value() == node) {
+                    portNum = p.index();
+                    break;
+                  }
+                FModuleOp referencedModule = dyn_cast<FModuleOp>(
+                    *instanceGraph.getReferencedModule(ins));
+                if (referencedModule) {
+                  auto port =
+                      referencedModule.getBodyBlock()->getArgument(portNum);
+                  // Get all the ports that have a comb path from `port`.
+                  for (const auto &portNode : portPaths[port])
+                    for (auto outputPort : portNode.getSecond()) {
+                      auto instResult = ins.getResult(
+                          outputPort.val.cast<BlockArgument>().getArgNumber());
+                      // Note that, in this case the instResult is always of
```

Comment:
Consider adding an `assert` for things like "this is something I expect to be true at this point".

Reply:
Thanks, but used the comment for clarity, Didn't add the assert, since this should be true by construction, `portPaths` is always updated accordingly.

Reply:
It's not a big deal here, but I have to respectfully object a little to "this should be true by construction" as a reason to *not* assert things, or "I am sure I update this accordingly" -- it makes this more fragile to changes by folks /not/ aware of such a requirement (or yourself in X time or under duress) and if this were something more reusable it'd be even more important to assert it.

Asserts for this sort of thing are great, IMO: they often document the property about as neatly as a comment and sometimes more so, ensure it's at least usually true in testing, very debugging-friendly, and are less easy to go stale unnoticed.
And they have basically no cost, honestly they're not very expensive to begin with but for situations where that's too much they're disabled (default distribution configuration, etc.).

See also:
https://llvm.org/docs/CodingStandards.html#assert-liberally .

This particular property I don't feel strongly about (and such a property if important maybe would be checked more centrally/consistently, or when consuming the data or something?), but I do about leveraging asserts haha :smile: .  Thanks for considering it regardless, I'll dismount my soap box now.

Reply:
Thanks @dtzSiFive for the soap box :) 
I realized this is important, added the assert in https://github.com/llvm/circt/pull/4521